Lib/pathlib.py

```diff
@@ -1193,18 +1236,254 @@ def cwd(cls):
         # os.path.abspath('.') == os.getcwd().
         return cls().absolute()
 
+    def expanduser(self):
+        """ Return a new path with expanded ~ and ~user constructs
+        (as returned by os.path.expanduser)
+        """
+        self._unsupported("expanduser")
+
     @classmethod
     def home(cls):
-        """Return a new path pointing to the user's home directory (as
-        returned by os.path.expanduser('~')).
+        """Return a new path pointing to expanduser('~').
         """
         return cls("~").expanduser()
 
+    def readlink(self):
+        """
+        Return the path to which the symbolic link points.
+        """
+        self._unsupported("readlink")
+
+    def resolve(self, strict=False):
+        """
+        Make the path absolute, resolving all symlinks on the way and also
+        normalizing it.
+        """
+        if self._resolving:
+            return self
+        try:
+            path = self.absolute()
+        except UnsupportedOperation:
+            path = self
+
+        def split(path):
+            return path._from_parsed_parts(path.drive, path.root, []), path._tail[::-1]
+
+        link_count = 0
+        stat_cache = {}
+        target_cache = {}
+        path, parts = split(path)
+        while parts:
+            part = parts.pop()
+            if part == '..':
+                if not path._tail:
+                    if path.root:
+                        # Delete '..' segment immediately following root
+                        continue
+                elif path._tail[-1] != '..':
+                    # Delete '..' segment and its predecessor
+                    path = path.parent
+                    continue
+                path = path._make_child_relpath(part)
+            else:
+                lookup_path = path
+                path = path._make_child_relpath(part)
+                path._resolving = True
+                path_str = str(path)
```

Comment:
Perhaps add a note on why we use the string representation here for the cache keys rather than using e.g. ``_drv, _root, _tail``?

Reply:
I'll switch to using the path object itself as a cache key. In future we might set `__hash__ = None` in `PathBase` and force users to write their own `__hash__()` if they want their path objects to be hashable, at which point using the `str(path)` as a cache key will be easier to justify.